### PR TITLE
Updated readme

### PR DESCRIPTION
**Pull request by mdabrowski1990**
# Description

- update other packages description (be more precise)
- add a few details about py-uds

# How Has This Been Tested?

- just review

# Process

I know the [process](https://github.com/mdabrowski1990/uds/wiki/Software-Development-Life-Cycle) and did my best to follow it

### CODE REVIEW

**Comment by mdabrowski1990** [comment]
@pylessard I would appreciate if you could review these changes.

File: README.rst [hunk 1]
```diff
@@ -55,26 +57,46 @@ python-udsoncan
...
-    - only Client side communication can be handled
-    - limited communication parameters configuration
+
+  - does not support simulating negatives scenarios on Transport/Network layer, e.g.
+
+    - cannot force Overflow / Wait value of Flow Status (parameter of Flow Control packet)
+    - cannot send CAN packets in wrong order (e.g. incorrect Sequence Numbers order in Consecutive Frames)
+
+  - only Client side communication is implemented - https://udsoncan.readthedocs.io/en/latest/udsoncan/client.html#
+  - no control or measurement of CAN timing parameters (N_As, N_Ar, N_Bs, N_Br, N_Cs, N_Cr)
```
Comment (pylessard) [inline README.rst:81]: These timings are ISOTP values. Not UDS.   My isotp package has control over timing, they're just given meaningful names.

File: README.rst [hunk 2]
```diff
@@ -55,26 +57,46 @@ python-udsoncan
...
+    https://udsoncan.readthedocs.io/en/latest/udsoncan/services.html
+  - positive and negatives scenarios are handled - https://udsoncan.readthedocs.io/en/latest/udsoncan/exceptions.html
+
 - cons:
-    - only positive use case scenarios (where communication is in line with UDS standard) are supported
-    - only CAN bus is supported
-    - only Client side communication can be handled
-    - limited communication parameters configuration
+
+  - does not support simulating negatives scenarios on Transport/Network layer, e.g.
```
Comment (pylessard) [inline README.rst:75]: Actually :  https://udsoncan.readthedocs.io/en/latest/udsoncan/client.html#overriding-the-output
It's not a super power feature, but it is enough to cover most negative tests

**Comment by pylessard** [comment]
Hey, I put some comments. 
Most of your comments revolves around ISOTP stuff that are simply not the responsability of the UDS layer. I made my project 100% decoupled from the underneath layer while I see that your project also includes ISOTP. Each approach has pros and cons, but it is wrong to assume that it is "missing". 

Modularity is always a bit more complex and harder to optimize. That could be the differentiator you put ahead : a turn-key solution that covers everything from application to hardware layer.

**Comment by pylessard** [comment]
Also, if you want to do something better than my repo, I suggest designing right from the start the possibility to have async calls with the client. My repo uses a stateful communication that can only accomodate a request/response approach. It's simple, but it has limitations. The nicer way would be to have a service thread that manages the communication. When a user "sends" a request, he is given a future object that can be awaited now or later.  Having a background task would also allow to register callbacks on certain broadcast messages, like  ResponseOnEvent

**Comment by mdabrowski1990** [comment]
> Also, if you want to do something better than my repo, I suggest designing right from the start the possibility to have async calls with the client. My repo uses a stateful communication that can only accomodate a request/response approach. It's simple, but it has limitations. The nicer way would be to have a service thread that manages the communication. When a user "sends" a request, he is given a future object that can be awaited now or later. Having a background task would also allow to register callbacks on certain broadcast messages, like ResponseOnEvent

Yeah, exactly that is the plan to have an option to do stuff asynchronously. I have not prepared Client/Server/Sniffer classes yet to handle these top layers (in Model OSI, it would be layer 5 - Session).
I tried to document idea here (https://uds.readthedocs.io/en/stable/pages/knowledge_base/osi_model.html) in case you are interested. So layers 1/2 are hopefully handled by other packages (python-can, CANlib and scapy would definately be supported, so far only python-can has been integrated), layers 3/4 would be part of this package (as you mention it has its pros and cons, but I decided that would be easier to integrate everything this way. Later on I might add option to limit your installation (for example only UDS, UDS + DoCAN or UDS + DoIP) to tackle some cons.

File: README.rst (hunk 1 above)
Comment (mdabrowski1990) [reply README.rst]: updated
please, have a look again

File: README.rst (hunk 2 above)
Comment (mdabrowski1990) [reply README.rst]: updated
please, have a look again

**Comment by mdabrowski1990** [comment]
@pylessard I would appreciate if you referred to knowledge base (it is documentation with explanation of how UDS works) in your project(s) and took part in reviewing whenever a new part is released. No pressure of course ;)

Link: https://uds.readthedocs.io/en/stable/pages/knowledge_base.html

**Comment by pylessard** [comment]
Thanks,
I agree more with what I read now :)
I can add the link without issue.

One last thing, just like this conversation, the knowledge base mixes UDS and ISOTP like they were the same thing. There is no such thing as a UDS packet. Segmentation is defined by ISOTP (iso-15765), not UDS. That is very misleading. There are UDS messages that can be sent over an ISOTP frame. ISOTP frames are segmented in many PDU. A PDU is expected to be a CAN message.

I see you are putting lot of effort to explain the protocol. That's good, but make sure to refer to the standard and be exact

**Review by pylessard — APPROVED** [review]

**Comment by pylessard** [comment]
https://udsoncan.readthedocs.io/en/latest/udsoncan/questions_answers.html#where-can-i-find-more-information-about-uds

**Pull request merged**